### PR TITLE
API Combine Sortable, Filterable and Limitable into SS_List

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/8778

1. Remove interfaces
- Sortable
- Filterable
- Limitable

2. Adds methods to SS_List
- columnUnique()
- count()
- excludeAny()
- exists()
- getIterator()
- offsetExists()
- offsetGet()
- offsetSet()
- offsetUnset()

3. Update ListDecorator
- Add excludeAny()
- Rename Count() to count()

4. Adds strong typing to all SS_List methods and implementing subclass methods

### CODE REVIEW

File: src/Model/List/ArrayList.php
Comment (GuySartorelli) [inline src/Model/List/ArrayList.php:565]: It is _theoretically_ possible to use `ints` as keys in `ArrayList` - but I'm okay with explicitly disallowing that. It _will_ need to be called out in the changelog though.

File: src/Model/List/ArrayList.php
Comment (GuySartorelli) [inline src/Model/List/ArrayList.php]: Unlike with `DataList`, ID could be anything, since any object/assoc array can be stored in an `ArrayList`.
```suggestion
    public function byID(mixed $id): mixed
```

File: src/Model/List/ListDecorator.php
Comment (GuySartorelli) [inline src/Model/List/ListDecorator.php:14]: Don't remove this! This represents the underlying list. It was correct, just need to remove the `&Sortable`etc from the end.

File: src/Model/List/ListDecorator.php
Comment (GuySartorelli) [inline src/Model/List/ListDecorator.php]: Revert this and any other reference to `TList`

File: src/Model/List/ListDecorator.php
Comment (GuySartorelli) [inline src/Model/List/ListDecorator.php]:
```suggestion
        return $this->count() > 0;
```
A list "exists" if it has values.

File: src/Model/List/ListDecorator.php
Comment (GuySartorelli) [inline src/Model/List/ListDecorator.php]:
```suggestion
    public function byID(mixed $id): mixed
```
Just like with `ArrayList`, we don't know what type of list it is and therefore we don't know what it contains so we don't know what values it might allow for an "ID"

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [inline src/Model/List/SS_List.php]:
```suggestion
```
This is part of the `Countable` interface. We don't need to double require it.

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [inline src/Model/List/SS_List.php:98]: If we're including this, then remove the conditional check on `ListDecorator` for it. Though I don't think we need to include it personally.

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [inline src/Model/List/SS_List.php]: I don't think these methods belong on a generic list interface.

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [inline src/Model/List/SS_List.php]: This is part of the `IteratorAggregate` interface already
```suggestion
```

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [inline src/Model/List/SS_List.php]:
```suggestion
```
These (and `offsetUnset()` below) are part of `ArrayAccess` already.

File: src/ORM/DataList.php
Comment (GuySartorelli) [inline src/ORM/DataList.php]:
```suggestion
    public function filterByCallback(callable $callback): SS_List
```
Lets allow ourselves to change that in the future. From an API perspective devs only need to know it has the `SSList` set of methods.
This lets us potentially return a `DataList` in the future for example, if we find a way to do that.

File: src/ORM/DataList.php
Comment (GuySartorelli) [inline src/ORM/DataList.php]:
```suggestion
```
That's just not needed at all.

File: src/ORM/DataList.php
Comment (GuySartorelli) [inline src/ORM/DataList.php]: How come this is mixed, but `first()` is specific? Looks like it should be:
```suggestion
    public function last(): ?DataObject
```

File: src/ORM/DataList.php
Comment (GuySartorelli) [inline src/ORM/DataList.php]: Why allow string? Is this for the edge case where someone may have replaced primary key functionality to allow UUID?

Also why allow null? If you want a list where the ID of `DataObject` records is null, you just want an empty list, which is kinda pointless.

File: src/ORM/DataList.php
Comment (GuySartorelli) [inline src/ORM/DataList.php:1907]: I just noticed `add()` doesn't even do anything so should probably throw a `BadMethodCallException`.... but that's technically out of scope so I'll let you decide if you want to do that.

File: src/ORM/EagerLoadedList.php
Comment (GuySartorelli) [inline src/ORM/EagerLoadedList.php:276]:
```suggestion
    public function remove(mixed $item): static
```
It's returning a value, so we should type that value.

Arguably this type should be added to all `add()` and `remove()` methods, so lists are more consistently "fluent".

At the very least they _do_ need to all have a return type added since you're adding it to some. It should be `void` if they return nothing or the type that they return if the do return something. IMO they should return consistently across the board but it's up to you if that's in scope or not.

File: src/ORM/DataList.php
Comment (GuySartorelli) [inline src/ORM/DataList.php]:
```suggestion
    public function find(string $key, mixed $value): ?DataObject
```

File: src/ORM/EagerLoadedList.php
Comment (GuySartorelli) [inline src/ORM/EagerLoadedList.php]:
```suggestion
    public function filterByCallback(callable $callback): SSList
```

File: src/ORM/ManyManyList.php
Comment (GuySartorelli) [inline src/ORM/ManyManyList.php]: Just don't allow null if null isn't valid. Do the same with `ManyManyThroughList` and anywhere else that has this.

File: src/ORM/Search/BasicSearchContext.php
Comment (GuySartorelli) [inline src/ORM/Search/BasicSearchContext.php:42]: Dunno why you changed from `instanceof` to `is_a` .... but since we have no policy about which to use I'll just pretend I didn't see that.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Model/List/ArrayList.php
Comment (emteknetnz) [reply src/Model/List/ArrayList.php]: OK, though this is for the $id, why would you use something that's not an int or a string as an $id?  It should always be 

i.e. This method calls `$this->filter("ID", $id)->first();`

File: src/Model/List/ListDecorator.php
Comment (emteknetnz) [reply src/Model/List/ListDecorator.php:14]: Updated

File: src/Model/List/ListDecorator.php
Comment (emteknetnz) [reply src/Model/List/ListDecorator.php]: Reverted

File: src/Model/List/ListDecorator.php
Comment (emteknetnz) [reply src/Model/List/ListDecorator.php]: Updated

File: src/Model/List/ListDecorator.php
Comment (emteknetnz) [reply src/Model/List/ListDecorator.php]: We know that it's an SS_List, which is type hinted for `int|string|null $id`?

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php]: Removed

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php:98]: Removed the check in ListDecorator

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php]: They're from Filterable, purpose of the card was to merge all the of these interfaces together

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php]: Removed

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php]: Removed

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php]: Removed

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php]: Updated

Also updated the same return type on ListDecorator::filterByCallback() and EagerLoadedList::filterByCallback()

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php]: Updated

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php]: Updated

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php]: Being pragmatic, running this through kitchen-sink there was of existing code that will pass in integers as strings, and null.

I put an inline comment on SS_List as to why I did this

It should by only int, including on SS_List, there's a docblock type of $int on Filterable::byID()

While we could fix up all our own references to this, it's a heavily used method and will break a lot of project s e.g. `$id = $request->param('ID');` will probably be a string

I'd by OK removing the `int|string|null` typehint on SS_List and just having it as an docblock $int instead

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php:1907]: Leave as is

File: src/ORM/EagerLoadedList.php
Comment (emteknetnz) [reply src/ORM/EagerLoadedList.php:276]: The DataList subclasses are inconsistent with whether they return a DataObject or void.  DataList returns void, HasManyList returns void, ManyManyList returns a DataObject

The other SS_List implementations are inconsistent too. EagerLoadedList returns itself, ArrayList returns void

I feel that it's better to just leave the return type blank across the board here as that's what's most consistent

File: src/ORM/EagerLoadedList.php
Comment (emteknetnz) [reply src/ORM/EagerLoadedList.php]: Updated

File: src/ORM/ManyManyList.php
Comment (emteknetnz) [reply src/ORM/ManyManyList.php]: Updated

File: src/Model/List/ArrayList.php
Comment (emteknetnz) [reply src/Model/List/ArrayList.php:565]: Have updated changelog PR

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [reply src/Model/List/SS_List.php]: I think we can allow ourselves to exclude methods that don't make sense in a generic list interface. They didn't make sense in a generic filterable interface either.

Excluding them also allows us to include appropriate typings for the various classes that implement this interface, as pointed out in various comments in this PR.

That said, to avoid holding onto this PR forever, I'll not hold up merging if you decide for some reason these methods _are_ appropriate for a generic list interface and choose to keep them.

File: src/ORM/EagerLoadedList.php
Comment (GuySartorelli) [reply src/ORM/EagerLoadedList.php:276]: We're adding explicit strong typing to these methods - we should consistently add it to the `remove()` method on _all_ classes. Though as you mention we can also leave the return type blank instead.

So we have three ways forward:
1. add the type that is currently used (`void` and `static` respectively depending on whether a value is returned)
2. consistently returning `$this` for all implementations of `remove()` and using `static` as the return typehint.
3. revert the changes that add typehints, though that feels like we're just ignoring a problem

I'm not terribly fussed about which you choose to do (though I do prefer 2 because it gives a more consistent developer experience) - but we should do one of these.

> I feel that it's better to just leave the return type blank across the board here as that's what's most consistent

The most consistent thing to do would be to always return `$this` so if we're aiming for consistency lets go with 2 above.

File: src/Model/List/SS_List.php
Comment (GuySartorelli) [inline src/Model/List/SS_List.php]: This comment placement is weird - it would make sense if a method was implemented here and this was a comment just before some code that normalises the value of that parameter, but that's not the case here.

Ideally the parameter typing should just be clear, and not need a lengthy comment to explain it.
I think I'd prefer just _not_ adding a type here, instead of having this lengthy comment.
If you insist on having a type that requires a lengthy comment, the comment should probably be in the `@param` docblock (which, btw, is currently incorrect in this method, as it says `int` which doesn't match the actual typehint).

File: src/ORM/DataList.php
Comment (GuySartorelli) [reply src/ORM/DataList.php]: The point of strong typing is to clean up inconsistencies like that.
If we're adding a strong typehint here, we should clean up the relevant core code so that it passes the correct typed value into the method.

The alternative is to just not include a type here, rather than saying "it can be a string or null or an int but should be int but isn't always int"

File: src/Model/List/ArrayList.php
Comment (GuySartorelli) [reply src/Model/List/ArrayList.php]: This is for a list which can contain any values for any reason. It should be `mixed` here because "ID" isn't something we control here. It could be an object of a `Uuid` class, for example.

File: src/Model/List/ListDecorator.php
Comment (GuySartorelli) [reply src/Model/List/ListDecorator.php]: Conversation can be held in https://github.com/silverstripe/silverstripe-framework/pull/11433#discussion_r1815652124 unless the change in https://github.com/silverstripe/silverstripe-framework/pull/11433#discussion_r1815668115 results in reverting it to not be strongly typed.

File: src/Model/List/ArrayList.php
Comment (emteknetnz) [reply src/Model/List/ArrayList.php]: Changed to mixed

File: src/Model/List/ListDecorator.php
Comment (emteknetnz) [reply src/Model/List/ListDecorator.php]: Changed to mixed

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php]: ArrayList would be the most "agnostic" list we have, meaning it's the list the least assumes you have DataObjects like objects in it.

However even that still has this:

```php
    public function byIDs($ids)
    {
        $ids = array_map('intval', $ids ?? []); // sanitize
        return $this->filter('ID', $ids);
    }
```

i.e. when calling the method it's assuming there's an integer 'ID' "column" available to query.  Obviously this method won't work if you fill the list with something without and ID column, though that's not really going to cause any real-world problems. Developers will get an exception if they call the method and adjust accordingly.

I think we just keep as is

Comment (GuySartorelli) [reply src/Model/List/SS_List.php]: > However even that still has this:

That's because a) our predecessors haven't always treated it as agnostically as they perhaps should have, and added some DataObject specific methods to it, and b) it implemented `Filterable` and therefore _had_ to have some implementation for that method.

In any case, as I said, to avoid holding onto this PR forever, I'll not hold up merging if you decide for some reason these methods are appropriate for a generic list interface and choose to keep them.

File: src/ORM/DataList.php
Comment (emteknetnz) [reply src/ORM/DataList.php]: Changed to mixed

File: src/ORM/EagerLoadedList.php
Comment (emteknetnz) [reply src/ORM/EagerLoadedList.php:276]: I don't want to change the the implemented return type remove() methods, particularly since EagerLoadedList returns static while ManyManyList returns a DataObject.  For this reason I want to keep this as is.

I think that's out of scope for this PR. I'm trying my best to constrain scope in PRs these days :-)

While I've added strong typing to this PR, that's respecting existing implementations.

The only implementation changes this PR has made have been:
- Throwing exceptions if methods don't exist on the inner list on ListDecorator for - forTemplate(), columnUnique(), and debug() - as those methods aren't defined on SS_List
- Refactoring func_get_args() to ...$args, which is functionality identical

File: src/Model/List/SS_List.php
Comment (emteknetnz) [reply src/Model/List/SS_List.php]: Have removed and changed $id type to mixed instead

File: src/ORM/EagerLoadedList.php
Comment (GuySartorelli) [reply src/ORM/EagerLoadedList.php:276]: That's fine, do 1 or 3 then.

File: src/ORM/DataList.php
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Shouldn't be mixed for DataList but whatever

File: src/ORM/EagerLoadedList.php
Comment (GuySartorelli) [reply src/ORM/EagerLoadedList.php:276]: Feels bad that we are _removing_ strict types in this PR but whatever at this point.

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**
